### PR TITLE
chore: Update CI workflow versions to use latest checkout (removes deprecation warning)

### PR DESCRIPTION
**Pull request by bryantbiggs**
- `actions/checkout` v3 uses deprecated nodejs runtime, no-op bump to v4 to remove warning
- Use python 3.9 as minimum version in CI workflows given 3.8 is coming to EOL
- Use Ubuntu 22.04 as minimum version in CI workflows given 18.04 is EOL
- Use Windows 2022 versions in CI workflows given 2019 is coming to EOL

### CODE REVIEW

File: .github/workflows/build-magma-linux.yml [hunk 1]
```diff
@@ -30,10 +30,10 @@ jobs:
     runs-on: linux.2xlarge
     strategy:
       matrix:
-        cuda_version: ["124", "121", "118"]
+        cuda_version: ["118", "121", "124"]
```
Comment (nWEIdia) [inline .github/workflows/build-magma-linux.yml:33]: What is the benefit of changing the order?

File: .github/workflows/build-manywheel-images.yml [hunk 2]
```diff
@@ -37,15 +37,15 @@ jobs:
     runs-on: linux.12xlarge.ephemeral
     strategy:
       matrix:
-        cuda_version: ["12.1", "11.8"]
+        cuda_version: ["11.8", "12.1", "12.4"]
```
Comment (nWEIdia) [inline .github/workflows/build-manywheel-images.yml:40]: Please see https://github.com/pytorch/builder/pull/1726

File: .github/workflows/build-nvidia-images.yml [hunk 3]
```diff
@@ -23,14 +23,14 @@ env:
   IN_CI: 1
   IS_GHA: 1
   IMAGE_NAME: "ghcr.io/pytorch/nvidia/cuda"
-  CUDA_VERSION: "11.5.0"
-  OS: "ubuntu1804"
-  OS_BASE: "ubuntu18.04"
+  CUDA_VERSION: "11.8.0"
+  OS: "ubuntu2204"
+  OS_BASE: "ubuntu22.04"
```
Comment (nWEIdia) [inline .github/workflows/build-nvidia-images.yml:28]: Good point!

File: .github/workflows/build-magma-linux.yml (hunk 1 above)
Comment (bryantbiggs) [reply .github/workflows/build-magma-linux.yml:33]: None! it was just trying to be consistent with increasing versions left-to-right

File: .github/workflows/build-manywheel-images.yml (hunk 2 above)
Comment (bryantbiggs) [reply .github/workflows/build-manywheel-images.yml]: thank you! reverted this change in https://github.com/pytorch/builder/pull/1734/commits/e5c34cadf79b956f5116a7a5e107f6c9ab207df2

File: .github/workflows/build-magma-windows.yml [hunk 4]
```diff
@@ -14,17 +14,17 @@ on:
 
 jobs:
   build-windows-magma:
-    runs-on: windows-2019
+    runs-on: windows-2022
     strategy:
       matrix:
-        cuda_version: ["121", "118"]
+        cuda_version: ["118", "121", "124"]
```
[the remark this answers is not in the document]
Comment (bryantbiggs) [reply .github/workflows/build-magma-windows.yml]: thank you! reverted this change in https://github.com/pytorch/builder/pull/1734/commits/e5c34cadf79b956f5116a7a5e107f6c9ab207df2

**Review by seemethere — APPROVED** [review]

**Comment by bryantbiggs** [comment]
hmmm https://github.com/actions/runner/issues/2906

**Comment by atalman** [comment]
@bryantbiggs this PR looks Ok, however there are some errors we want to resolve:
```
/home/ec2-user/actions-runner/externals/node20/bin/node: /lib64/libm.so.6: version `GLIBC_2.27' not found (required by /home/ec2-user/actions-runner/externals/node20/bin/node)
```

https://github.com/pytorch/builder/actions/runs/8970189807/job/24636329217?pr=1734

**Comment by bryantbiggs** [comment]
@atalman correct - I opened this #1762 that I believe would need to be resolved first, but I think only internal Meta employees would be able to handle that

**Pull request closed without merging**